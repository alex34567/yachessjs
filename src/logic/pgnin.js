'use strict'

const pieces = require('./pieces')
const state = require('./state')
const moveNotation = require('./move_notation')

const PGN_TOKEN_REGEX = /([[\].*()<>]|"(?:[^\\"]|\\"|\\\\)+"|$\d+|[A-Za-z0-9][A-Za-z0-9_+#=:\-!?/]*)|{.*?}|;.*$|^%.*|\s+/my
const TAGNAME_REGEX = /[a-zA-Z0-9_]+/

class PGNGame {
  constructor (tags, state) {
    this.tags = tags
    this.state = state
  }
}

function parsePgn (pgnStr) {
  const db = []
  const tokenStream = pgnStr.split(new RegExp(PGN_TOKEN_REGEX)).filter(x => x)
  let tokenIndex = 0

  while (tokenIndex < tokenStream.length) {
    const tags = []
    while (tokenStream[tokenIndex] === '[') {
      tokenIndex++
      const name = tokenStream[tokenIndex]
      tokenIndex++
      if (!TAGNAME_REGEX.test(name)) {
        throw new Error('PGN Parse Error')
      }
      const rawValue = tokenStream[tokenIndex]
      tokenIndex++
      if (rawValue[0] !== '"') {
        throw new Error('PGN never ending string')
      }
      const value = rawValue.substring(1, rawValue.length - 1)
      tags.push([name, value])
      if (tokenStream[tokenIndex] !== ']') {
        throw new Error('PGN Parse Error')
      }
      tokenIndex++
    }
    const setupIndex = tags.findIndex(x => x[0] === 'SetUp')
    const fenIndex = tags.findIndex(x => x[0] === 'FEN')
    let gameState = state.START_STATE
    if (setupIndex >= 0 && tags[setupIndex][1] === '1' && fenIndex >= 0) {
      gameState = state.stateFromFen(tags[fenIndex][1])
    }
    while (true) {
      tokenIndex++
      if (/^[0-9]+$/.test(tokenStream[tokenIndex - 1])) {
        if (String(gameState.moveCount()) !== tokenStream[tokenIndex - 1]) {
          throw new Error('Invalid game number enumeration')
        }
        continue
      }
      if (tokenStream[tokenIndex - 1][0] === '$') {
        continue
      }
      let done = false
      switch (tokenStream[tokenIndex - 1]) {
        case '(': {
          let depth = 1
          while (depth > 0) {
            if (tokenStream[tokenIndex] === '(') {
              depth++
            } else if (tokenStream[tokenIndex] === ')') {
              depth--
            }
            tokenIndex++
          }
          continue
        }
        case '.':
          continue
        case '1-0':
          if (gameState.currTurn !== pieces.BLACK) {
            throw new Error('Invalid result')
          }
          if (!gameState.isCheckmate()) {
            gameState = gameState.modify(x => { x.resign = true })
          }
          done = true
          break
        case '0-1':
          if (gameState.currTurn !== pieces.WHITE) {
            throw new Error('Invalid result')
          }
          if (!gameState.isCheckmate()) {
            gameState = gameState.modify(x => { x.resign = true })
          }
          done = true
          break
        case '1/2-1/2':
          if (!gameState.isDraw()) {
            gameState = gameState.modify(x => { x.agreeDraw = true })
          }
          done = true
          break
        case '*':
          done = true
          break
      }
      if (done) {
        break
      }
      const move = tokenStream[tokenIndex - 1]
      const parsedMove = moveNotation.MoveNotation.parseMove(gameState.currTurn, move)
      if (!parsedMove) {
        throw new Error('Move syntax error')
      }
      const possibleMoves = gameState.moves().filter(x => !x.invalid() && parsedMove.matches(x))
      if (possibleMoves.length === 0) {
        throw new Error('Invalid move tried to be played')
      }
      if (possibleMoves.length > 1) {
        throw new Error('Move ambiguity in PGN file')
      }
      gameState = possibleMoves[0].do()
    }
    db.push(new PGNGame(tags, gameState))
  }

  return db
}

Object.assign(module.exports, { parsePgn })
