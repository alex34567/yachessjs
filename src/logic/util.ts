function convertFileLetter (letter: string) {
  const charCode = letter.toLowerCase().charCodeAt(0)
  return charCode - 'a'.charCodeAt(0)
}

function convertRankLetter (letter: string) {
  return Number(letter) - 1
}

class Pos {
  file: number;
  rank: number;

  constructor (file: number, rank: number) {
    this.file = file
    this.rank = rank

    if (!Number.isInteger(rank) || !Number.isInteger(file) || rank < 0 || rank > 7 || file < 0 || file > 7) {
      throw new Error('Invalid board pos')
    }
    // Object.freeze(this)
  }

  // Returns null on out of bounds
  add (file: number, rank: number) {
    const newFile = this.file + file
    const newRank = this.rank + rank
    if (newFile < 0 || newFile > 7 || newRank < 0 || newRank > 7) {
      return null
    }
    return new Pos(newFile, newRank)
  }

  addFile (file: number) {
    return this.add(file, 0)
  }

  addRank (rank: number) {
    return this.add(0, rank)
  }

  toRaw () {
    return this.rank * 8 + this.file
  }

  static fromRaw (raw: number): Pos {
    const rank = Math.floor(raw / 8)
    const file = raw % 8
    return new Pos(file, rank)
  }

  toString () {
    const fileLetter = String.fromCharCode(this.file + 'a'.charCodeAt(0))
    const rankLetter = String(this.rank + 1)
    return fileLetter + rankLetter
  }

  compare (other: Pos) {
    const fileCompare = this.file - other.file
    if (fileCompare !== 0) {
      return fileCompare
    }
    return this.rank - other.rank
  }
}

class PgnMoveLineGen {
  currMoveStr: string;
  totalMoveStr: string;

  constructor () {
    this.currMoveStr = ''
    this.totalMoveStr = ''
  }

  addToken (token: string) {
    if (this.currMoveStr === '') {
      this.currMoveStr = token
    } else {
      const tmpMoveStr = this.currMoveStr + ' ' + token
      if (tmpMoveStr.length > 80) {
        this.totalMoveStr += '\n' + this.currMoveStr
        this.currMoveStr = token
      } else {
        this.currMoveStr = tmpMoveStr
      }
    }
  }

  done () {
    return this.totalMoveStr + '\n' + this.currMoveStr
  }
}

export {
  convertFileLetter,
  convertRankLetter,
  Pos,
  PgnMoveLineGen
}
