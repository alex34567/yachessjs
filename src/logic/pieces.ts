import * as moves from './moves'
import * as state from './state'
import * as util from './util'
import * as immutable from 'immutable'
import { State } from './state'

class ConstructorKey {}

abstract class Square {
  constructor (_key: ConstructorKey) {
    if (typeof this.isOccupied !== 'function') {
      throw new Error('isOccupied function not defined')
    }

    if (typeof this.canMoveOnto !== 'function') {
      throw new Error('canMoveOnto function not defined')
    }

    if (typeof this.canBeCaptured !== 'function') {
      throw new Error('canBeCaptured function not defined')
    }
  }

  id = 0
  abstract isOccupied (): this is Piece;
  abstract canMoveOnto (piece: Piece): boolean;
  abstract canBeCaptured (piece: Piece): boolean;
}

class EmptySquare extends Square {
  isOccupied () {
    return false
  }

  canMoveOnto (_other: Piece) {
    return true
  }

  canBeCaptured (_other: Piece) {
    return false
  }
}

function moveLine (state: state.State, piece: Piece, fromPos: util.Pos, line: [number, number]) {
  const moveList = []
  let toPos = fromPos.add(line[0], line[1])
  while (toPos && state.board.get(toPos).canMoveOnto(piece)) {
    moveList.push(new moves.NormalMove(state, piece, fromPos, toPos))
    // Once the piece hits another piece stop
    if (state.board.get(toPos).canBeCaptured(piece)) {
      break
    }
    toPos = toPos.add(line[0], line[1])
  }
  return moveList
}

function protectLine (state: state.State, piece: Piece, fromPos: util.Pos, line: [number, number]) {
  const protects = []
  let toPos = fromPos.add(line[0], line[1])
  while (toPos && state.board.get(toPos).canMoveOnto(piece)) {
    protects.push(toPos)
    // Once the piece hits another piece stop
    if (state.board.get(toPos).canBeCaptured(piece)) {
      break
    }
    toPos = toPos.add(line[0], line[1])
  }

  if (toPos) {
    protects.push(toPos)
  }
  return protects
}

function applyCheckPinFlag (state: state.State, fromPos: util.Pos, line: [number, number]) {
  let toPos = fromPos.add(line[0], line[1])
  while (toPos && state.board.get(toPos) !== state.currTurn.KING) {
    state.board.setPinned(toPos, true)
    toPos = toPos.add(line[0], line[1])
  }
}

function checkForKing (state: state.State, piece: Piece, fromPos: util.Pos, line: [number, number]) {
  let toPos = fromPos.add(line[0], line[1])
  while (toPos && state.board.get(toPos).canMoveOnto(piece)) {
    if (state.board.get(toPos) === state.currTurn.KING) {
      return true
    }
    if (state.board.get(toPos).canBeCaptured(piece)) {
      break
    }
    toPos = toPos.add(line[0], line[1])
  }
  return false
}

function pinLine (state: state.State, piece: Piece, fromPos: util.Pos, line: [number, number]) {
  let toPos = fromPos.add(line[0], line[1])
  while (toPos && state.board.get(toPos).canMoveOnto(piece)) {
    if (state.board.get(toPos).canBeCaptured(piece)) {
      if (state.board.get(toPos) === state.currTurn.KING) {
        applyCheckPinFlag(state, fromPos, line)
      } else if (checkForKing(state, piece, toPos, line)) {
        state.board.setPinned(toPos, true)
        let pinnedAxis: 0 | 1 | 2 | 3 = 0
        if ((line[0] === 1 && line[1] === 0) || (line[0] === -1 && line[1] === 0)) {
          pinnedAxis = 1
        } else if ((line[0] === -1 && line[1] === 1) || (line[0] === 1 && line[1] === -1)) {
          pinnedAxis = 2
        } else if ((line[0] === -1 && line[1] === -1) || (line[0] === 1 && line[1] === 1)) {
          pinnedAxis = 3
        }
        state.board.setPinnedAxis(toPos, pinnedAxis)
      }
      break
    }
    toPos = toPos.add(line[0], line[1])
  }
}

abstract class Piece extends Square {
  color: Readonly<Color>
  fenLetter: string

  constructor (key: ConstructorKey, color: Readonly<Color>, fenLetter: string, id: number) {
    super(key)

    if (typeof this.moves !== 'function') {
      throw new Error('Moves function not defined')
    }
    if (typeof this.getPGNLetter !== 'function') {
      throw new Error('Get pgn letter not defined')
    }

    this.color = color
    this.fenLetter = fenLetter
    this.id = id
  }

  isOccupied () {
    return true
  }

  canMoveOnto (other: Piece) {
    return this.color !== other.color
  }

  canBeCaptured (other: Piece) {
    return this.canMoveOnto(other)
  }

  // This method will get a mutable state
  pin (state: State, pos: util.Pos) {}

  abstract moves (state: state.State, myPos: util.Pos): moves.Move[]
  abstract protects (state: state.State, myPos: util.Pos): util.Pos[]
  abstract getPGNLetter(): string
}

class Pawn extends Piece {
  addMove (state: state.State, moveList: moves.Move[], fromPos: util.Pos, toPos: util.Pos) {
    if (toPos.rank === this.color.OTHER_COLOR.KING_RANK) {
      for (const choice of this.color.PROMOTE_LIST) {
        moveList.push(new moves.Promotion(state, this, fromPos, toPos, choice))
      }
    } else {
      moveList.push(new moves.NormalMove(state, this, fromPos, toPos))
    }
  }

  getPGNLetter () {
    return ''
  }

  moves (state: state.State, myPos: util.Pos) {
    const moveList: moves.Move[] = []
    const forwardMove = myPos.addRank(this.color.PAWN_RANK_DIR)

    if (!forwardMove) {
      return moveList
    }
    const forwardOccupied = state.board.get(forwardMove!).isOccupied()
    if (!forwardOccupied) {
      this.addMove(state, moveList, myPos, forwardMove!)
    }
    if (!forwardOccupied && this.color.PAWN_RANK === myPos.rank) {
      const firstMove = myPos.addRank(this.color.PAWN_RANK_DIR * 2)
      console.assert(firstMove, 'First move should be in bounds')
      if (!state.board.get(firstMove!).isOccupied()) {
        moveList.push(new moves.FirstPawn(state, this, myPos, firstMove!))
      }
    }
    const otherColor = state.getColor(this.color.OTHER_COLOR)
    const leftDiag = myPos.add(-1, this.color.PAWN_RANK_DIR)

    if (leftDiag && state.board.get(leftDiag).canBeCaptured(this)) {
      this.addMove(state, moveList, myPos, leftDiag)
    }
    const rightDiag = myPos.add(1, this.color.PAWN_RANK_DIR)
    if (rightDiag && state.board.get(rightDiag).canBeCaptured(this)) {
      this.addMove(state, moveList, myPos, rightDiag)
    }
    if (otherColor.enPassantPos) {
      if (leftDiag && otherColor.enPassantPos.compare(leftDiag) === 0) {
        moveList.push(new moves.EnPassant(state, this, myPos, leftDiag))
      }
      if (rightDiag && otherColor.enPassantPos.compare(rightDiag) === 0) {
        moveList.push(new moves.EnPassant(state, this, myPos, rightDiag))
      }
    }

    return moveList
  }

  protects (state: State, myPos: util.Pos): util.Pos[] {
    const protects = []
    const leftDiag = myPos.add(-1, this.color.PAWN_RANK_DIR)
    if (leftDiag) {
      protects.push(leftDiag)
    }
    const rightDiag = myPos.add(1, this.color.PAWN_RANK_DIR)
    if (rightDiag) {
      protects.push(rightDiag)
    }
    return protects
  }
}

class Rook extends Piece {
  moves (state: state.State, pos: util.Pos): moves.Move[] {
    const line: [number, number][] = [[0, 1], [0, -1], [1, 0], [-1, 0]]
    return (line
      .map(line => moveLine(state, this, pos, line))
      .reduce((acc, val) => acc.concat(val)))
  }

  getPGNLetter () {
    return 'R'
  }

  pin (state: State, pos: util.Pos) {
    const line: [number, number][] = [[0, 1], [0, -1], [1, 0], [-1, 0]]
    return line.forEach(line => pinLine(state, this, pos, line))
  }

  protects (state: State, myPos: util.Pos): util.Pos[] {
    const line: [number, number][] = [[0, 1], [0, -1], [1, 0], [-1, 0]]
    return (line
      .map(line => protectLine(state, this, myPos, line))
      .reduce((acc, val) => acc.concat(val)))
  }
}

class Knight extends Piece {
  moves (state: state.State, pos: util.Pos): moves.Move[] {
    const knightPos = [[1, 2], [-1, 2], [1, -2], [-1, -2], [2, 1], [-2, 1], [2, -1], [-2, -1]]
    const moveList = []
    for (const deltaPos of knightPos) {
      const newPos = pos.add(deltaPos[0], deltaPos[1])
      if (newPos && state.board.get(newPos).canMoveOnto(this)) {
        moveList.push(new moves.NormalMove(state, this, pos, newPos))
      }
    }
    return moveList
  }

  protects (state: State, pos: util.Pos): util.Pos[] {
    const knightPos = [[1, 2], [-1, 2], [1, -2], [-1, -2], [2, 1], [-2, 1], [2, -1], [-2, -1]]
    const protects = []
    for (const deltaPos of knightPos) {
      const newPos = pos.add(deltaPos[0], deltaPos[1])
      if (newPos) {
        protects.push(newPos)
      }
    }
    return protects
  }

  getPGNLetter () {
    return 'N'
  }
}

class Bishop extends Piece {
  moves (state: state.State, pos: util.Pos): moves.Move[] {
    const line: [number, number][] = [[1, 1], [-1, -1], [1, -1], [-1, 1]]
    return (line
      .map(line => moveLine(state, this, pos, line))
      .reduce((acc, val) => acc.concat(val)))
  }

  getPGNLetter () {
    return 'B'
  }

  pin (state: State, pos: util.Pos) {
    const line: [number, number][] = [[1, 1], [-1, -1], [1, -1], [-1, 1]]
    return line.forEach(line => pinLine(state, this, pos, line))
  }

  protects (state: State, myPos: util.Pos): util.Pos[] {
    const line: [number, number][] = [[1, 1], [-1, -1], [1, -1], [-1, 1]]
    return (line
      .map(line => protectLine(state, this, myPos, line))
      .reduce((acc, val) => acc.concat(val)))
  }
}

class Queen extends Piece {
  moves (state: state.State, pos: util.Pos) {
    return Bishop.prototype.moves.call(this, state, pos).concat(Rook.prototype.moves.call(this, state, pos))
  }

  getPGNLetter () {
    return 'Q'
  }

  pin (state: state.State, pos: util.Pos) {
    Bishop.prototype.pin.call(this, state, pos)
    Rook.prototype.pin.call(this, state, pos)
  }

  protects (state: State, pos: util.Pos): util.Pos[] {
    return Bishop.prototype.protects.call(this, state, pos).concat(Rook.prototype.protects.call(this, state, pos))
  }
}

class King extends Piece {
  moves (state: state.State, pos: util.Pos): moves.Move[] {
    const kingPos = [[1, 0], [1, 1], [-1, 0], [-1, -1], [1, -1], [-1, 1], [0, 1], [0, -1]]
    const moveList = []
    for (const deltaPos of kingPos) {
      const newPos = pos.add(deltaPos[0], deltaPos[1])
      if (newPos && state.board.get(newPos).canMoveOnto(this)) {
        moveList.push(new moves.NormalMove(state, this, pos, newPos))
      }
    }
    return moveList
  }

  protects (state: State, pos: util.Pos): util.Pos[] {
    const kingPos = [[1, 0], [1, 1], [-1, 0], [-1, -1], [1, -1], [-1, 1], [0, 1], [0, -1]]
    const protects = []
    for (const deltaPos of kingPos) {
      const newPos = pos.add(deltaPos[0], deltaPos[1])
      if (newPos) {
        protects.push(newPos)
      }
    }
    return protects
  }

  getPGNLetter () {
    return 'K'
  }
}

export interface Color {
  PAWN: Readonly<Piece>,
  ROOK: Readonly<Piece>,
  KNIGHT: Readonly<Piece>,
  BISHOP: Readonly<Piece>,
  QUEEN: Readonly<Piece>,
  KING: Readonly<Piece>,
  KING_RANK: number,
  PAWN_RANK: number,
  PAWN_RANK_DIR: number,
  PROMOTE_LIST: Readonly<Readonly<Piece>[]>
  FROM_LETTER: immutable.Map<string, Readonly<Piece>>,
  OTHER_COLOR: Readonly<Color>,
}

function genColor (kingRank: number, pawnRank: number, pawnRankDir: number, fenConv: (l: string) => string, idOffset: number) {
  const ret = {} as Color
  const key = new ConstructorKey()
  Object.assign(ret, {
    PAWN: Object.freeze(new Pawn(key, ret, fenConv('P'), 1 + idOffset)),
    KNIGHT: Object.freeze(new Knight(key, ret, fenConv('N'), 2 + idOffset)),
    BISHOP: Object.freeze(new Bishop(key, ret, fenConv('B'), 3 + idOffset)),
    ROOK: Object.freeze(new Rook(key, ret, fenConv('R'), 4 + idOffset)),
    QUEEN: Object.freeze(new Queen(key, ret, fenConv('Q'), 5 + idOffset)),
    KING: Object.freeze(new King(key, ret, fenConv('K'), 6 + idOffset)),
    KING_RANK: kingRank,
    PAWN_RANK: pawnRank,
    PAWN_RANK_DIR: pawnRankDir
  })
  ret.PROMOTE_LIST = Object.freeze([ret.QUEEN, ret.KNIGHT, ret.ROOK, ret.BISHOP])
  const fromLetter: [string, Piece][] = [['K', ret.KING], ['Q', ret.QUEEN], ['R', ret.ROOK],
    ['B', ret.BISHOP], ['N', ret.KNIGHT], ['', ret.PAWN]]
  ret.FROM_LETTER = immutable.Map(fromLetter)
  return ret
}

const WRITABLE_WHITE = genColor(0, 1, 1, x => x, 0)
const WRITABLE_BLACK = genColor(7, 6, -1, x => x.toLowerCase(), 8)
WRITABLE_WHITE.OTHER_COLOR = WRITABLE_BLACK
Object.freeze(WRITABLE_WHITE)
WRITABLE_BLACK.OTHER_COLOR = WRITABLE_WHITE
Object.freeze(WRITABLE_BLACK)

const WHITE: Readonly<Color> = WRITABLE_WHITE
const BLACK: Readonly<Color> = WRITABLE_BLACK

const FROM_FEN_ARRAY: [string, Readonly<Piece>][] = [['K', WHITE.KING], ['Q', WHITE.QUEEN], ['R', WHITE.ROOK],
  ['B', WHITE.BISHOP], ['N', WHITE.KNIGHT], ['P', WHITE.PAWN],
  ['k', BLACK.KING], ['q', BLACK.QUEEN], ['r', BLACK.ROOK],
  ['b', BLACK.BISHOP], ['n', BLACK.KNIGHT], ['p', BLACK.PAWN]]

const FROM_FEN = immutable.Map(FROM_FEN_ARRAY)

const EMPTY = Object.freeze(new EmptySquare(new ConstructorKey()))

export {
  EMPTY,
  WHITE,
  BLACK,
  FROM_FEN,
  Piece,
  Square
}
