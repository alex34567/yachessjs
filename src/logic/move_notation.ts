import * as util from './util'
import * as moves from './moves'
import * as pieces from './pieces'

const NORMAL_MOVE_REGEX = /^([KQRBN]?)([a-h]?)([1-8]?)(x?)([a-h])([1-8])(?:=?([QRBN]))?(?:\+|#|)(?:\?\??|!!?|\?!|!\?|)$/

abstract class MoveNotation {
  protected constructor () {
    if (typeof this.matches !== 'function') {
      throw new Error('Matches function not defined')
    }
  }

  static parseMove (color: pieces.Color, move: string) {
    move = move.trim()
    if (move === 'O-O' || move === '0-0') {
      return new CastleMoveNotation(true)
    }
    if (move === 'O-O-O' || move === '0-0-0') {
      return new CastleMoveNotation(false)
    }
    const parsedMove = NORMAL_MOVE_REGEX.exec(move)
    if (!parsedMove) {
      return null
    }

    const pieceLetter = parsedMove[1]
    const piece = color.FROM_LETTER.get(pieceLetter)!
    const fromFileLetter = parsedMove[2]
    const fromFile = fromFileLetter ? util.convertFileLetter(fromFileLetter) : null
    const fromRankLetter = parsedMove[3]
    const fromRank = fromRankLetter ? util.convertRankLetter(fromRankLetter) : null
    const captures = parsedMove[4] === 'x'
    const toFile = util.convertFileLetter(parsedMove[5])
    const toRank = util.convertRankLetter(parsedMove[6])
    const toPos = new util.Pos(toFile, toRank)
    const promoteLetter = parsedMove[7]
    let promoteChoice = null
    if (promoteLetter) {
      promoteChoice = color.FROM_LETTER.get(promoteLetter)!
    }
    return new NormalMoveNotation(piece, captures, fromFile, fromRank, toPos, promoteChoice)
  }

  abstract matches (move: moves.Move): boolean
}

class CastleMoveNotation extends MoveNotation {
  isKingSide: boolean

  constructor (isKingSide: boolean) {
    super()
    this.isKingSide = isKingSide
  }

  toString () {
    if (this.isKingSide) {
      return 'O-O'
    } else {
      return 'O-O-O'
    }
  }

  matches (move: moves.Move): boolean {
    return move.isCastle() && this.isKingSide === move.isKingSide
  }
}

class NormalMoveNotation extends MoveNotation {
  piece: pieces.Piece
  captures: boolean
  fromFile: number | null
  fromRank: number | null
  toPos: util.Pos
  promoteChoice: pieces.Piece | null
  checkSymbol: string

  constructor (piece: pieces.Piece, captures: boolean, fromFile: number | null, fromRank: number | null, toPos: util.Pos, promoteChoice: pieces.Piece | null) {
    super()
    this.piece = piece
    this.captures = captures
    this.fromFile = fromFile
    this.fromRank = fromRank
    this.toPos = toPos
    this.promoteChoice = promoteChoice
    this.checkSymbol = ''
  }

  toString () {
    let captureLetter = ''
    if (this.captures) {
      captureLetter = 'x'
    }
    let fromFileLetter = ''
    if (this.fromFile !== null) {
      fromFileLetter = String.fromCharCode(this.fromFile + 'a'.charCodeAt(0))
    }
    let fromRankDigit = ''
    if (this.fromRank !== null) {
      fromRankDigit = String(this.fromRank + 1)
    }
    let promoteStr = ''
    if (this.promoteChoice) {
      promoteStr += '=' + this.promoteChoice.getPGNLetter()
    }
    return this.piece.getPGNLetter() + fromFileLetter + fromRankDigit + captureLetter + this.toPos.toString() + promoteStr + this.checkSymbol
  }

  matches (move: moves.Move): boolean {
    // Promotes are only valid for pawns
    if (this.piece !== move.state.currTurn.PAWN && this.promoteChoice) {
      return false
    }
    if (!move.isNormal()) {
      return false
    }
    // Be permissive in allowing a capture when not specified but require that a move like Qxe6 is a capture
    if (this.captures && !(move.state.board.get(this.toPos).canBeCaptured(move.piece) || move.isEnPassant())) {
      return false
    }
    if (move.isPromote() && this.promoteChoice !== move.promoteChoice) {
      return false
    }
    if (!move.isPromote() && this.promoteChoice) {
      return false
    }
    return move.piece === this.piece &&
            (this.fromFile === null || this.fromFile === move.fromPos.file) &&
            (this.fromRank === null || this.fromRank === move.fromPos.rank) &&
            move.toPos.compare(this.toPos) === 0
  }
}

export { CastleMoveNotation, MoveNotation, NormalMoveNotation }
