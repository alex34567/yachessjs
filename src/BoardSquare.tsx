import { BLACK, Square, WHITE } from './logic/pieces'
import * as React from 'react'
import { Theme, ThemeManager } from './theme'

export interface BoardSquareProps {
  isBlack?: boolean
  inCheck?: boolean
  piece: Square
  onClick?: () => void
  highlighted?: boolean
  canMoveTo?: boolean
  children?: React.ReactNode
  theme: ThemeManager
}

function getPieceName (piece: Square): string | undefined {
  switch (piece) {
    case WHITE.QUEEN:
      return 'White Queen'
    case WHITE.KING:
      return 'White King'
    case WHITE.BISHOP:
      return 'White Bishop'
    case WHITE.ROOK:
      return 'White Rook'
    case WHITE.KNIGHT:
      return 'White Knight'
    case WHITE.PAWN:
      return 'White Pawn'
    case BLACK.QUEEN:
      return 'Black Queen'
    case BLACK.KING:
      return 'Black King'
    case BLACK.BISHOP:
      return 'Black Bishop'
    case BLACK.ROOK:
      return 'Black Rook'
    case BLACK.KNIGHT:
      return 'Black Knight'
    case BLACK.PAWN:
      return 'Black Pawn'
  }
  return undefined
}

function getPieceImage (theme: Theme, piece: Square): React.ReactNode | undefined {
  const name = getPieceName(piece)
  if (name) {
    return <img alt={name} className='ChessPiece' src={process.env.PUBLIC_URL + `pieces/${theme.piece.prefix}/${name}.svg`} />
  }
  return undefined
}

export default function BoardSquare (props: BoardSquareProps) {
  let className = `ChessBoardSquare ${props.theme.className}`
  if (props.isBlack) {
    className += ' ChessBoardSquareBlack'
  } else {
    className += ' ChessBoardSquareWhite'
  }
  let highlight
  if (props.highlighted) {
    highlight =
      <svg className='Overlay'>
        <rect width='100%' height='100%' fill='yellow' fillOpacity='.5' />
      </svg>
  }
  let checkHighlight
  if (props.inCheck) {
    checkHighlight =
      <svg className='Overlay'>
        <rect width='100%' height='100%' fill='red' fillOpacity='.5' />
      </svg>
  }
  const pieceImage = getPieceImage(props.theme.theme, props.piece)
  let moveIndicator
  if (props.canMoveTo && !pieceImage) {
    moveIndicator =
      <svg className='Overlay'>
        <circle r='10%' cx='50%' cy='50%' />
      </svg>
  }
  if (props.canMoveTo && pieceImage) {
    moveIndicator =
      <svg className='Overlay'>
        <circle r='47.5%' cx='50%' cy='50%' fill='none' stroke='black' strokeWidth='2.5%' />
      </svg>
  }
  return (
    <div className={className} onClick={props.onClick}>
      {highlight}
      {checkHighlight}
      {pieceImage}
      {moveIndicator}
      {props.children}
    </div>
  )
}
