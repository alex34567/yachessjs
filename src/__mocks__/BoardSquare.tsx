import React from 'react'
import { BoardSquareProps } from '../BoardSquare'

function falseToUndef (x: unknown): true | undefined {
  if (!x) {
    return
  }
  return true
}

export default function BoardSquare (props: BoardSquareProps) {
  let name
  if (props.piece.isOccupied()) {
    name = props.piece.fenLetter
  }
  return (
    <div
      data-name={name}
      data-isblack={falseToUndef(props.isBlack)}
      data-incheck={falseToUndef(props.inCheck)}
      data-canmoveto={falseToUndef(props.canMoveTo)}
      data-highlighted={falseToUndef(props.highlighted)}
      onClick={props.onClick}
    >
      {props.children}
    </div>
  )
}
