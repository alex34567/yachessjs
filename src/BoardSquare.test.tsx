import { act } from '@testing-library/react'
import BoardSquare from './BoardSquare'
import * as Pieces from './logic/pieces'
import React from 'react'
import { render, unmountComponentAtNode } from 'react-dom'

// Setup code adapted react docs
let container: HTMLDivElement | null = null
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement('div')
  document.body.appendChild(container)
})

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container!)
  container!.remove()
  container = null
})

test('Empty Square', () => {
  act(() => {
    render(<BoardSquare piece={Pieces.EMPTY} />, container)
  })
  expect(container).toMatchSnapshot()
})

test('Empty Black Square', () => {
  act(() => {
    render(<BoardSquare piece={Pieces.EMPTY} isBlack />, container)
  })
  expect(container).toMatchSnapshot()
})

test('White Pawn', () => {
  act(() => {
    render(<BoardSquare piece={Pieces.WHITE.PAWN} />, container)
  })
  expect(container).toMatchSnapshot()
})

test('Selected Queen', () => {
  act(() => {
    render(<BoardSquare piece={Pieces.BLACK.QUEEN} highlighted />, container)
  })
  expect(container).toMatchSnapshot()
})

test('In check', () => {
  act(() => {
    render(<BoardSquare piece={Pieces.BLACK.KING} inCheck />, container)
  })
  expect(container).toMatchSnapshot()
})

test('Highlighted in check', () => {
  act(() => {
    render(
      <BoardSquare piece={Pieces.BLACK.KING} inCheck highlighted />,
      container
    )
  })
  expect(container).toMatchSnapshot()
})

test('Can Move To', () => {
  act(() => {
    render(<BoardSquare piece={Pieces.EMPTY} canMoveTo />, container)
  })
  expect(container).toMatchSnapshot()
})

test('Can capture', () => {
  act(() => {
    render(<BoardSquare piece={Pieces.WHITE.PAWN} canMoveTo />, container)
  })
  expect(container).toMatchSnapshot()
})

test('Clickable', () => {
  const click = jest.fn()
  act(() => {
    render(<BoardSquare onClick={click} piece={Pieces.EMPTY}/>, container)
  })

  const square = container?.firstElementChild
  expect(square).not.toBeNull()
  expect(click).not.toBeCalled()

  act(() => {
    square!.dispatchEvent(new MouseEvent('click', { bubbles: true }))
  })

  expect(click).toBeCalled()
})
