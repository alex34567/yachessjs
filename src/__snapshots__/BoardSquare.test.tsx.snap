// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Can Move To 1`] = `
<div>
  <div
    class="ChessBoardSquare TestTheme ChessBoardSquareWhite"
  >
    <div
      class="MoveDot"
    />
  </div>
</div>
`;

exports[`Can capture 1`] = `
<div>
  <div
    class="ChessBoardSquare TestTheme ChessBoardSquareWhite"
  >
    <img
      alt="White Pawn"
      class="ChessPiece"
      src="/pieces/cburnett/White Pawn.svg"
    />
    <svg
      class="Overlay"
    >
      <circle
        cx="50%"
        cy="50%"
        fill="none"
        r="47.5%"
        stroke="black"
        stroke-width="2.5%"
      />
    </svg>
  </div>
</div>
`;

exports[`Empty Black Square 1`] = `
<div>
  <div
    class="ChessBoardSquare TestTheme ChessBoardSquareBlack"
  />
</div>
`;

exports[`Empty Square 1`] = `
<div>
  <div
    class="ChessBoardSquare TestTheme ChessBoardSquareWhite"
  />
</div>
`;

exports[`Highlighted in check 1`] = `
<div>
  <div
    class="ChessBoardSquare TestTheme ChessBoardSquareWhite"
  >
    <div
      class="Highlight"
    />
    <div
      class="InCheck"
    />
    <img
      alt="Black King"
      class="ChessPiece"
      src="/pieces/cburnett/Black King.svg"
    />
  </div>
</div>
`;

exports[`In check 1`] = `
<div>
  <div
    class="ChessBoardSquare TestTheme ChessBoardSquareWhite"
  >
    <div
      class="InCheck"
    />
    <img
      alt="Black King"
      class="ChessPiece"
      src="/pieces/cburnett/Black King.svg"
    />
  </div>
</div>
`;

exports[`Selected Queen 1`] = `
<div>
  <div
    class="ChessBoardSquare TestTheme ChessBoardSquareWhite"
  >
    <div
      class="Highlight"
    />
    <img
      alt="Black Queen"
      class="ChessPiece"
      src="/pieces/cburnett/Black Queen.svg"
    />
  </div>
</div>
`;

exports[`White Pawn 1`] = `
<div>
  <div
    class="ChessBoardSquare TestTheme ChessBoardSquareWhite"
  >
    <img
      alt="White Pawn"
      class="ChessPiece"
      src="/pieces/cburnett/White Pawn.svg"
    />
  </div>
</div>
`;
